mllib/src/main/scala/org/apache/spark/mllib/feature/Word2Vec.scala

```diff
@@ -439,9 +439,21 @@ class Word2Vec extends Serializable with Logging {
           }
         }.flatten
       }
-      val synAgg = partial.reduceByKey { case (v1, v2) =>
-          blas.saxpy(vectorSize, 1.0f, v2, 1, v1, 1)
-          v1
+      // SPARK-24666: do normalization for aggregating weights from partitions.
+      // Original Word2Vec either single-thread or multi-thread which do Hogwild-style aggregation.
+      // Our approach needs to do extra normalization, otherwise adding weights continuously may
+      // cause overflow on float and lead to infinity/-infinity weights.
+      val synAgg = partial.mapPartitions { iter =>
+        iter.map { case (id, vec) =>
+          (id, (vec, 1))
+        }
+      }.reduceByKey { case ((v1, count1), (v2, count2)) =>
+        blas.saxpy(vectorSize, 1.0f, v2, 1, v1, 1)
+        (v1, count1 + count2)
+      }.map { case (id, (vec, count)) =>
+        val averagedVec = Array.fill[Float](vectorSize)(0.0f)
+        blas.saxpy(vectorSize, 1.0f / count, vec, 1, averagedVec, 1)
```

Comment:
Can this just use sscal, in place, or does it need a copy? it seems like it can just modify in place but I may be forgetting the semantics here.

Reply:
sscal will scales the vector in place.

Reply:
Yes right. I'm wondering if that's a good thing. Well, this is just done once for each aggregate so hardly matters.